### PR TITLE
Pod4 i2c

### PR DESCRIPTION
**Pull request by rgw3d**
Roses are Red, 
I2C is cool, 
It actually worked this time, 
I guess I'm not a fool.

This is a working pull request for someone else to clean up and finish implementing. I2C values can be read in from sensors -- the hard part is done. Now we need to schedule how often the reads will happen,  read in all of the values, store them, and return them from the `refresh()` function when called.

### CODE REVIEW

**Comment by rgw3d** [comment]
When reading the ADC values, note that there could be up to 4 actual ADCs (where they have addresses 0x49, 0x50...)

Each ADC then has 4 ports, ANC0, ANC1...

we possibly want to make the addresses/ number of ADCs configurable, but it's probably much much easier to just hardcode. It's not like it changes much

File: CentralComputing/I2CManager.cpp [hunk 1]
```diff
@@ -98,24 +98,38 @@ bool I2CManager::single_shot(int fd, int port, int16_t *value) {
...
+  if (i == 3) {
+    i = 0;
+  } else {
+    i++;
+  }
+  int port;
+  int addr;
+  if (i == 0) {
+    port = ANC0;
+    addr = 0x48;
```
Comment (rgw3d) [inline CentralComputing/I2CManager.cpp]: Please note my comment about the addresses.

Comment (anshulshah957) [reply CentralComputing/I2CManager.cpp]: I hardcoded it in for now. Want to get a test for the I2C up and running and focus on some priorities first. If we are still in good shape I can revisit making it more configurable

File: CentralComputing/I2CManager.cpp [hunk 2]
```diff
@@ -143,8 +143,8 @@ std::shared_ptr<I2CData> I2CManager::refresh() {
     int64_t b = Utils::microseconds(); 
     print(LogLevel::LOG_INFO, "i2c val: %d, in micros: %lu\n", value, b-a);
   }
-  
-  new_data->temp[i] = value;
+  int index = (j * 4) + i;
```
Comment (rgw3d) [inline CentralComputing/I2CManager.cpp:153]: Exactly the way to do it!

**Comment by rgw3d** [comment]
ADS1115 addresses:

0x48
0x49
0x4A
0x4B

Its hex!

Supported here: https://learn.adafruit.com/adafruit-4-channel-adc-breakouts/assembly-and-wiring

File: CentralComputing/I2CManager.cpp [hunk 3]
```diff
@@ -137,16 +140,25 @@ std::shared_ptr<I2CData> I2CManager::refresh() {
   int16_t value = 0;
   
   if (!set_i2c_addr(i2c_fd, addr)) {
-    return false;
+    //TODO: @Anshul, add error here.
```
Comment (rgw3d) [inline CentralComputing/I2CManager.cpp]: @anshulshah957 Add error generation here. It's not hard, But you should see how to do it.

File: CentralComputing/I2CManager.cpp [hunk 4]
```diff
@@ -137,16 +140,25 @@ std::shared_ptr<I2CData> I2CManager::refresh() {
...
+    return empty_data();
   }
   if (single_shot(i2c_fd, port, &value)) {
     int64_t b = Utils::microseconds(); 
-    print(LogLevel::LOG_INFO, "i2c val: %d, in micros: %lu\n", value, b-a);
+    print(LogLevel::LOG_DEBUG, "(REMOVE) I2C: i=%d, j =%d, val: %d, took this long: micros: %lu\n", i, j, value, b-a);
+  } else {
+    //TODO: @Anshul, add error here.
```
Comment (rgw3d) [inline CentralComputing/I2CManager.cpp]: Add Error generation here. its `set_error_flag()`

File: CentralComputing/I2CManager.h [hunk 5]
```diff
@@ -27,19 +27,15 @@ class I2CManager : public SourceManagerBase<I2CData> {
   int32_t error_general_1_over_temp;
   int32_t error_general_2_over_temp;
   int32_t error_general_3_over_temp;
+  std::shared_ptr<I2CData> old_data;
```
Comment (rgw3d) [inline CentralComputing/I2CManager.h:30]: Please note how I added `old_data` and how it is used

File: CentralComputing/I2CManager.cpp [hunk 6]
```diff
@@ -137,16 +140,25 @@ std::shared_ptr<I2CData> I2CManager::refresh() {
...
+    return empty_data();
   }
   int index = (j * 4) + i; 
-  new_data->temp[index] = value;
 
-  // myfloat = val * VPS; // convert to voltage
+  // Update the "old" data with the new reading
+  old_data->temp[index] = value;
```
Comment (rgw3d) [inline CentralComputing/I2CManager.cpp:156]: See how I update `old_data`, which is then copied into `new_data`

File: CentralComputing/I2CManager.cpp [hunk 7]
```diff
@@ -173,13 +173,13 @@ void I2CManager::check_for_sensor_error(const std::shared_ptr<I2CData> & check_d
   auto temp_arr = check_data->temp; 
   for (int index = 0; index < NUM_TMP; index++) {
     if (temp_arr[index] > (int) error_general_1_over_temp) {
-      // Send error 1 here
+      Command::set_error_flag(Command::SET_I2C_ERROR,I2C_OVER_TEMP_ONE);
```
Comment (rgw3d) [inline CentralComputing/I2CManager.cpp]: From my reading of this, it looks like only one of these errors could possible cause a trigger. The lowest "Error_general_X_over_temp"  will cause the error to be triggered.

Either have just 1 error, or check  different indexes with the different error conditions

**Review by rgw3d — COMMENTED** [review]
2 Problems with the error code checking:

1) Why three errors? Currently checking all 3 error levels against all data. Makes more sense to check a sub-set of each data against only 1 error. Maybe have 3 sub-sets of data? 


2) UNITS. The units right now are in ADC levels (2^16 levels). Which is fine because we don't know the conversion into celcius because we are missing details on the thermistors. But I reccomend that you DON"T change anything. Leverage the Configuration file to save work in the future. Have it such that the Error trigger is `20000`, or something like that, (we change it later based on future specifications) which isn't a temperature, but makes sense as an ADC level. That value might actually just refer to 3V, which might actually mean 40 degrees C. It makes sense for us to just deal with this value in terms of ADC level. More flexible, less code later. 
_However_, do note that the values we are sending back to the Basestation are still in ADC levels. That means the Basestation will have to convert them into meaningful values in order to display them.

File: CentralComputing/I2CManager.cpp (hunk 7 above)
Comment (anshulshah957) [reply CentralComputing/I2CManager.cpp]: My bad I thought that there were 3 ascending errors (like a REALLY HIGH temp would be worse than a slightly high one)

**Pull request merged**